### PR TITLE
105559652 fix commit sorting

### PR DESCRIPTION
**Pull request by konto-andrzeja**
# Description
- I decided against sorting commits in a query, in my opinion it's much more readable now
- Commits are sorted by priority based on their state (fixed and accepted commits have the same priority) and then by date based on what priority they have, results will be something like this:
```
old rejected commit
new rejected commit
old auto rejected commit
new auto rejected commit
old passed commit
new passed commit
pending commit expiring today
pending commit expiring tomorrow
new fixed commit
new accepted commit
old fixed commit
old accepted commit
```
- I wrote specs to be sure that commits are sorted as above

# Pivotal ticket:
https://www.pivotaltracker.com/story/show/105559652

# Sign-offs:
- [ ] @mamut

### CODE REVIEW

File: spec/controllers/api/v2/commits_controller_spec.rb [hunk 1]
```diff
@@ -7,7 +7,7 @@
            project_id: project.id,
            author_id: author.id,
            authored_at: Date.parse,
-           expires_at: Date.parse - 1.day,
+           expires_at: Date.tomorrow,
```
Comment (mamut) [inline spec/controllers/api/v2/commits_controller_spec.rb]: I wouldn't recommend running any time-dependent tests without the use of `Timecop` – https://github.com/travisjeffery/timecop

File: spec/controllers/api/v2/commits_controller_spec.rb [hunk 2]
```diff
@@ -67,15 +67,32 @@
...
+               author_id: author.id,
+               expires_at: Date.today,
+               state: 'pending')
+      end
+      %w(rejected auto_rejected passed fixed accepted).each do |state|
+        let!("new_#{state}_commit") do
+          create(:commit,
+                 project_id: project.id,
+                 author_id: author.id,
+                 created_at: Date.today,
```
Comment (mamut) [inline spec/controllers/api/v2/commits_controller_spec.rb]: rather than setting `created_at` directly better to use Timecop

using `Date.today` and `Date.tomorrow` without freezing the time with timecop makes it possible for your specs to fail when run around midnight ;) one might say this is an edge case, but it's important to keep your tests robust and dependable

File: spec/controllers/api/v2/commits_controller_spec.rb [hunk 3]
```diff
@@ -86,15 +103,37 @@
...
       end
     end
-   context 'with valid user teruns all but accepted and fixed' do
+
+    context 'with valid user returns all but accepted and fixed' do
+      let!(:commit_accepted) do
+        create(:commit,
+               project_id: project.id,
+               author_id: author.id,
+               authored_at: Date.parse - 2.days,
```
Comment (mamut) [inline spec/controllers/api/v2/commits_controller_spec.rb]: `Date.parse - 2.days` this is weird, didn't know you could call `Date.parse` without an argument. But it would be better to use something less exotic not to scare off any newcomers, like `2.days.ago` for example (which might've been your original intent anyway)

File: app/controllers/api/v2/commits_controller.rb [hunk 4]
```diff
@@ -69,8 +69,26 @@ def commit_pages_count
...
+
+  def compare_same_priority_commits(x, y)
+    case x.state.to_sym
+    when :accepted, :fixed
+      y.created_at <=> x.created_at
+    when :pending
+      y.expires_at <=> x.expires_at
+    else
+      x.created_at <=> y.created_at
+    end
```
Comment (mamut) [inline app/controllers/api/v2/commits_controller.rb]: I'm not entirely convinced this should be a part of the controller – it does specify the display order in this action, so it feels right in the controller, but at the same time the logic seems so low-level that I'd consider moving this to the model

tough choice

**Comment by mamut** [comment]
Not enitrely sold on the in-memory sorting, but let's go with the simpler solution and revisit this when we actually run into any performance issues ;)

File: app/controllers/api/v2/commits_controller.rb [hunk 5]
```diff
@@ -69,8 +69,26 @@ def commit_pages_count
...
+  end
+
+  def compare_same_priority_commits(x, y)
+    case x.state.to_sym
+    when :accepted, :fixed
+      y.created_at <=> x.created_at
+    when :pending
+      y.expires_at <=> x.expires_at
+    else
+      x.created_at <=> y.created_at
```
Comment (mamut) [inline app/controllers/api/v2/commits_controller.rb]: now you only need to cover two cases really, when `state == :pending` and everything else

File: spec/controllers/api/v2/commits_controller_spec.rb [hunk 6]
```diff
@@ -86,15 +103,37 @@
...
+                                                    old_auto_rejected_commit.id,
+                                                    new_auto_rejected_commit.id,
+                                                    old_passed_commit.id,
+                                                    new_passed_commit.id,
+                                                    pending_commit_expiring_today.id,
+                                                    pending_commit_expiring_tomorrow.id,
+                                                    new_fixed_commit.id,
+                                                    old_fixed_commit.id,
+                                                    new_accepted_commit.id,
+                                                    old_accepted_commit.id
```
Comment (mamut) [inline spec/controllers/api/v2/commits_controller_spec.rb]: i think it makes more sense to mix the fixed and accepted commits together – right now when you go to the list of all commits you need to scroll through all fixed commits before you can see the accepted ones, even if they were accepted a minute ago and the fixes are already a month old ;)

File: spec/controllers/api/v2/commits_controller_spec.rb (hunk 3 above)
Comment (konto-andrzeja) [inline spec/controllers/api/v2/commits_controller_spec.rb]: @mamut That's not my code, I just moved it and left it as it was, since this weird `Date.parse` without arguments is used in other places as well. I tried not to fix anything not related to what I did, since it apparently worked for some time. Maybe it's some hack, I don't know. Should I fix those anyway?

File: app/controllers/api/v2/commits_controller.rb (hunk 5 above)
Comment (konto-andrzeja) [inline app/controllers/api/v2/commits_controller.rb]: @mamut Nope, please take a closer look at those - `y.created_at <=> x.created_at` is not the same as `x.created_at <=> y.created_at`. The first one sorts in descending order, the second in ascending one.

File: spec/controllers/api/v2/commits_controller_spec.rb (hunk 1 above)
Comment (konto-andrzeja) [inline spec/controllers/api/v2/commits_controller_spec.rb]: @mamut I didn't know about that gem. Nice one, thanks!

File: app/controllers/api/v2/commits_controller.rb (hunk 4 above)
Comment (konto-andrzeja) [inline app/controllers/api/v2/commits_controller.rb]: @mamut I thought about it too and then just left it in the controller. When it's not entirely clear where to put something, best to leave it where it is. :)

**Comment by mamut** [comment]
looking good
